cloudinit/user_data.py

```diff
@@ -78,66 +85,100 @@ def __init__(self, paths):
         self.paths = paths
         self.ssl_details = util.fetch_ssl_details(paths)
 
-    def process(self, blob):
+    def process(self, blob, require_signature=False):
         accumulating_msg = MIMEMultipart()
         if isinstance(blob, list):
             for b in blob:
-                self._process_msg(convert_string(b), accumulating_msg)
+                self._process_msg(
+                    convert_string(b), accumulating_msg, require_signature
+                )
         else:
-            self._process_msg(convert_string(blob), accumulating_msg)
+            self._process_msg(
+                convert_string(blob), accumulating_msg, require_signature
+            )
         return accumulating_msg
 
-    def _process_msg(self, base_msg, append_msg):
+    def _process_msg(
+        self, base_msg: Message, append_msg, require_signature=False
+    ):
         def find_ctype(payload):
             return handlers.type_from_starts_with(payload)
 
         for part in base_msg.walk():
             if is_skippable(part):
                 continue
 
-            ctype = None
-            ctype_orig = part.get_content_type()
             payload = util.fully_decoded_payload(part)
-            was_compressed = False
 
-            # When the message states it is of a gzipped content type ensure
-            # that we attempt to decode said payload so that the decompressed
-            # data can be examined (instead of the compressed data).
-            if ctype_orig in DECOMP_TYPES:
-                try:
-                    payload = util.decomp_gzip(payload, quiet=False)
-                    # At this point we don't know what the content-type is
-                    # since we just decompressed it.
-                    ctype_orig = None
-                    was_compressed = True
-                except util.DecompressionError as e:
-                    error_message = (
-                        "Failed decompressing payload from {} of"
-                        " length {} due to: {}".format(
-                            ctype_orig, len(payload), e
-                        )
-                    )
-                    _handle_error(error_message, e)
-                    continue
+            ctype = part.get_content_type()
 
-            # Attempt to figure out the payloads content-type
-            if not ctype_orig:
-                ctype_orig = UNDEF_TYPE
             # There are known cases where mime-type text/x-shellscript included
             # non shell-script content that was user-data instead.  It is safe
             # to check the true MIME type for x-shellscript type since all
```

Comment:
> If that's what this is saying, then this could be a bug, because the "shebang is required for shell scripts" assumption is untrue

Is it? Trying to pass a mime message of type x-shellscript with script having no shebang leads to a cc_scripts_user failure due to a subp failure:
```
ProcessExecutionError("Exec format error. Missing #! in script?\nCommand: ['/var/lib/cloud/instance/scripts/test.sh']\nExit code: -\nReason: [Errno 8] Exec format error: b'/var/lib/cloud/instance/scripts/test.sh'\nStdout: -\nStderr: -")
```

I think http://www.faqs.org/faqs/unix-faq/faq/part3/section-16.html explains why.

Reply:
> > If that's what this is saying, then this could be a bug, because the "shebang is required for shell scripts" assumption is untrue
> 
> Is it? Trying to pass a mime message made from `cloud-init devel make-mime -a test.sh:x-shellscript` with script having no shebang leads to a cc_scripts_user failure due to a subp failure:
> 
> ```
> ProcessExecutionError("Exec format error. Missing #! in script?\nCommand: ['/var/lib/cloud/instance/scripts/test.sh']\nExit code: -\nReason: [Errno 8] Exec format error: b'/var/lib/cloud/instance/scripts/test.sh'\nStdout: -\nStderr: -")
> ```
> 
> I think http://www.faqs.org/faqs/unix-faq/faq/part3/section-16.html explains why.

No you're right. That link describes csh, but dash's man page describes this behavior as well. This is a thing that we could support if we wanted to with a subshell, but I don't think we need to.

I still have questions about what "known cases where mime-type text/x-shellscript included shell-script content that was user-data instead". This sounds like a potential bug.